### PR TITLE
AG-7905 - Fix listener handling on chart type changes

### PR DESCRIPTION
**Pull request by alantreadway**
https://ag-grid.atlassian.net/browse/AG-7905

Fixes setting of `listeners` and `legend.listeners` using `AgChart.updateDelta()` for Integrated Charts use-cases.

Additionally adds a new example to demonstrate this.

### CODE REVIEW

File: charts-packages/ag-charts-community/src/chart/agChartV2.ts [hunk 1]
```diff
@@ -353,20 +353,20 @@ function debug(message?: any, ...optionalParams: any[]): void {
 }
 
 function applyChartOptions(chart: Chart, processedOptions: Partial<AgChartOptions>, userOptions: AgChartOptions): void {
+    let skip = ['type', 'data', 'series', 'autoSize', 'listeners', 'theme', 'legend.listeners'];
```
Comment (alexanderby) [inline charts-packages/ag-charts-community/src/chart/agChartV2.ts:356]: That's it: `legend.listeners`

File: charts-packages/ag-charts-community/src/chart/agChartV2.ts [hunk 2]
```diff
@@ -395,7 +395,7 @@ function applyChartOptions(chart: Chart, processedOptions: Partial<AgChartOption
         registerListeners(chart, processedOptions.listeners);
     }
     if (processedOptions.legend?.listeners) {
-        Object.assign(chart.legend.listeners, processedOptions.legend.listeners);
+        Object.assign(chart.legend.listeners, processedOptions.legend.listeners ?? {});
```
Comment (alexanderby) [inline charts-packages/ag-charts-community/src/chart/agChartV2.ts:398]: Curious how is `undefined` treated here

File: charts-packages/ag-charts-community/src/util/observable.ts [hunk 3]
```diff
@@ -8,6 +8,10 @@ export class Observable {
     private allEventListeners = new Map<string, Set<TypedEventListener>>();
 
     addEventListener(type: string, listener: TypedEventListener): void {
+        if (typeof listener !== 'function') {
+            throw new Error('AG Charts - listener must be a Function');
```
Comment (alexanderby) [inline charts-packages/ag-charts-community/src/util/observable.ts:12]: A warning probably?

**Review by alexanderby — APPROVED** [review]

File: charts-packages/ag-charts-community/src/chart/agChartV2.ts (hunk 2 above)
Comment (alantreadway) [reply charts-packages/ag-charts-community/src/chart/agChartV2.ts:398]: LHS is never `undefined` (as it's out implementation model), but if it were then `Object.assign()` fails with:
```
> Object.assign(undefined, {})
Uncaught TypeError: Cannot convert undefined or null to object
    at Function.assign (<anonymous>)
```

Comment (alexanderby) [reply charts-packages/ag-charts-community/src/chart/agChartV2.ts:398]: I mean, what happened when the second argument was undefined?

File: charts-packages/ag-charts-community/src/util/observable.ts (hunk 3 above)
Comment (alantreadway) [reply charts-packages/ag-charts-community/src/util/observable.ts:12]: So I added a warning in the `AgChartsV2.ts` changes, and this method never gets called - this is just belt & braces guard against the internal state becoming corrupted if there are any other use-cases.

File: charts-packages/ag-charts-community/src/chart/agChartV2.ts (hunk 1 above)
Comment (alantreadway) [reply charts-packages/ag-charts-community/src/chart/agChartV2.ts:356]: Yes, the root problem before was that the `jsonDiff` output was resulting in `legend: { listeners: undefined }` (which is correct behaviour), and then this code was applying that `undefined` value, even though these was distinct handling later in this method :-(

Adding `legend: { listeners: {} }` to the base theme removes the `undefined` from the mix, and this addition means we're not trying to process the update in two different ways 🤞

File: charts-packages/ag-charts-community/src/chart/agChartV2.ts (hunk 2 above)
Comment (alantreadway) [reply charts-packages/ag-charts-community/src/chart/agChartV2.ts:398]:
```
> Object.assign({}, undefined)
{}
```

So its strictly unnecessary, but I feel that the addition of `?? {}` here at least clarifies the intent and doesn't rely on knowledge of the detail of `Object.assign()` to be readable.

**Pull request merged**